src/Symfony/Component/Messenger/Command/FailedMessagesRetryCommand.php

```diff
@@ -87,6 +87,8 @@ protected function configure(): void
      */
     protected function execute(InputInterface $input, OutputInterface $output)
     {
+        $this->eventDispatcher->addSubscriber(new StopWorkerOnMessageLimitListener(1));
```

Comment:
it feels a bit odd for us to mutate the event dispatcher... and also not clean up (remove these runtime event listeners)... but I can't think of any real reason this would be an issue.

Reply:
Could we pass the command input options to the `Worker` constructor and expose them, so that subscribers can be registered at compile time and adapt their behavior according to the worker options?

Reply:
@chalasr I don't see what you mean

Reply:
@Tobion let me try to illustrate with the `StopWorkerOnMessageLimitListener`:

```diff
diff --git a/src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml b/src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml
index 14117ee8e4..172d96752f 100644
--- a/src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml
+++ b/src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml
@@ -129,6 +129,12 @@
             <tag name="kernel.event_subscriber" />
         </service>
 
+        <service id="messenger.listener.stop_worker_on_message_limit_listener" class="Symfony\Component\Messenger\EventListener\StopWorkerOnMessageLimitListener">
+            <tag name="kernel.event_subscriber" />
+            <tag name="monolog.logger" channel="messenger" />
+            <argument type="service" id="logger" on-invalid="ignore" />
+        </service>
+
         <!-- routable message bus -->
         <service id="messenger.routable_message_bus" class="Symfony\Component\Messenger\RoutableMessageBus">
             <argument /> <!-- Message bus locator -->
diff --git a/src/Symfony/Component/Messenger/Command/ConsumeMessagesCommand.php b/src/Symfony/Component/Messenger/Command/ConsumeMessagesCommand.php
index 2347ce3c8c..a38c9efbab 100644
--- a/src/Symfony/Component/Messenger/Command/ConsumeMessagesCommand.php
+++ b/src/Symfony/Component/Messenger/Command/ConsumeMessagesCommand.php
@@ -197,7 +197,7 @@ EOF
 
         $bus = $input->getOption('bus') ? $this->routableBus->getMessageBus($input->getOption('bus')) : $this->routableBus;
 
-        $worker = new Worker($receivers, $bus, $this->eventDispatcher, $this->logger);
+        $worker = new Worker($receivers, $bus, $limit, $this->eventDispatcher, $this->logger);
         $worker->run([
             'sleep' => $input->getOption('sleep') * 1000000,
         ]);
diff --git a/src/Symfony/Component/Messenger/EventListener/StopWorkerOnMessageLimitListener.php b/src/Symfony/Component/Messenger/EventListener/StopWorkerOnMessageLimitListener.php
index ca71ff10bb..bc41d2797b 100644
--- a/src/Symfony/Component/Messenger/EventListener/StopWorkerOnMessageLimitListener.php
+++ b/src/Symfony/Component/Messenger/EventListener/StopWorkerOnMessageLimitListener.php
@@ -22,28 +22,22 @@ use Symfony\Component\Messenger\Exception\InvalidArgumentException;
  */
 class StopWorkerOnMessageLimitListener implements EventSubscriberInterface
 {
-    private $maximumNumberOfMessages;
     private $logger;
     private $receivedMessages = 0;
 
-    public function __construct(int $maximumNumberOfMessages, LoggerInterface $logger = null)
+    public function __construct(LoggerInterface $logger = null)
     {
-        $this->maximumNumberOfMessages = $maximumNumberOfMessages;
         $this->logger = $logger;
-
-        if ($maximumNumberOfMessages <= 0) {
-            throw new InvalidArgumentException('Message limit must be greater than zero.');
-        }
     }
 
     public function onWorkerRunning(WorkerRunningEvent $event): void
     {
-        if (!$event->isWorkerIdle() && ++$this->receivedMessages >= $this->maximumNumberOfMessages) {
+        if (!$event->isWorkerIdle() && ($maximumNumberOfMessages = $event->getWorker()->getMaximumNumberOfMessages()) && ++$this->receivedMessages >= $maximumNumberOfMessages) {
             $this->receivedMessages = 0;
             $event->getWorker()->stop();
 
             if (null !== $this->logger) {
-                $this->logger->info('Worker stopped due to maximum count of {count} messages processed', ['count' => $this->maximumNumberOfMessages]);
+                $this->logger->info('Worker stopped due to maximum count of {count} messages processed', ['count' => $maximumNumberOfMessages]);
             }
         }
     }
diff --git a/src/Symfony/Component/Messenger/Worker.php b/src/Symfony/Component/Messenger/Worker.php
index 546690efa5..d6837b60f0 100644
--- a/src/Symfony/Component/Messenger/Worker.php
+++ b/src/Symfony/Component/Messenger/Worker.php
@@ -38,15 +38,17 @@ class Worker
     private $bus;
     private $eventDispatcher;
     private $logger;
+    private $maximumNumberOfMessages;
     private $shouldStop = false;
 
     /**
      * @param ReceiverInterface[] $receivers Where the key is the transport name
      */
-    public function __construct(array $receivers, MessageBusInterface $bus, EventDispatcherInterface $eventDispatcher = null, LoggerInterface $logger = null)
+    public function __construct(array $receivers, MessageBusInterface $bus, int $maximumNumberOfMessages = null, EventDispatcherInterface $eventDispatcher = null, LoggerInterface $logger = null)
     {
         $this->receivers = $receivers;
         $this->bus = $bus;
+        $this->maximumNumberOfMessages = $maximumNumberOfMessages;
         $this->eventDispatcher = LegacyEventDispatcherProxy::decorate($eventDispatcher);
         $this->logger = $logger;
     }
@@ -150,6 +152,11 @@ class Worker
         $this->shouldStop = true;
     }
 
+    public function getMaximumNumberOfMessages(): ?int
+    {
+        return $this->maximumNumberOfMessages;
+    }
+
     private function dispatchEvent($event)
     {
         if (null === $this->eventDispatcher) {

```

makes sense?

Reply:
The StopWorkerOnMessageLimitListener would then always be active even if it's not abled. Also it seems the worker shouldn't know about maximumNumberOfMessages if it doesn't implement it.
So I prefer the current solution.